FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/pkcs11_operations.c

```diff
@@ -0,0 +1,1038 @@
+                                     const char * pcPrivateKey,
+                                     size_t xPrivateKeyLength,
+                                     const char * pcLabel )
+{
+    CK_RV xResult = CKR_OK;
+    mbedtls_pk_type_t xMbedKeyType = MBEDTLS_PK_NONE;
+    int xMbedResult = 0;
+    mbedtls_pk_context xMbedPkContext = { 0 };
+
+    mbedtls_pk_init( &xMbedPkContext );
+    xMbedResult = mbedtls_pk_parse_key( &xMbedPkContext, ( const uint8_t * ) pcPrivateKey,
+                                        xPrivateKeyLength, NULL, 0 );
+
+    if( xMbedResult != 0 )
+    {
+        LogError( ( "Unable to parse private key." ) );
+        xResult = CKR_ARGUMENTS_BAD;
+    }
+
+    /* Determine whether the key to be imported is RSA or EC. */
+    if( xResult == CKR_OK )
+    {
+        xMbedKeyType = mbedtls_pk_get_type( &xMbedPkContext );
+
+        if( xMbedKeyType == MBEDTLS_PK_RSA )
+        {
+            xResult = prvProvisionPrivateRSAKey( xSession, pcLabel, &xMbedPkContext );
+        }
+        else if( ( xMbedKeyType == MBEDTLS_PK_ECDSA ) ||
+                 ( xMbedKeyType == MBEDTLS_PK_ECKEY ) ||
+                 ( xMbedKeyType == MBEDTLS_PK_ECKEY_DH ) )
+        {
+            xResult = prvProvisionPrivateECKey( xSession, pcLabel, &xMbedPkContext );
+        }
+        else
+        {
+            LogError( ( "Invalid private key type provided. Only RSA-2048 and "
+                        "EC P-256 keys are supported." ) );
+            xResult = CKR_ARGUMENTS_BAD;
+        }
+    }
+
+    mbedtls_pk_free( &xMbedPkContext );
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV prvProvisionCertificate( CK_SESSION_HANDLE xSession,
+                                      const char * pcCertificate,
+                                      size_t xCertificateLength,
+                                      const char * pcLabel )
+{
+    PKCS11_CertificateTemplate_t xCertificateTemplate;
+    CK_OBJECT_CLASS xCertificateClass = CKO_CERTIFICATE;
+    CK_CERTIFICATE_TYPE xCertificateType = CKC_X_509;
+    CK_FUNCTION_LIST_PTR xFunctionList = NULL;
+    CK_RV xResult = CKR_OK;
+    uint8_t * pucDerObject = NULL;
+    int32_t ulConversion = 0;
+    size_t xDerLen = 0;
+    CK_BBOOL xTokenStorage = CK_TRUE;
+    CK_BYTE pxSubject[] = "TestSubject";
+    CK_OBJECT_HANDLE xObjectHandle = CK_INVALID_HANDLE;
+
+    /* Initialize the client certificate template. */
+    xCertificateTemplate.xObjectClass.type = CKA_CLASS;
+    xCertificateTemplate.xObjectClass.pValue = &xCertificateClass;
+    xCertificateTemplate.xObjectClass.ulValueLen = sizeof( xCertificateClass );
+    xCertificateTemplate.xSubject.type = CKA_SUBJECT;
+    xCertificateTemplate.xSubject.pValue = pxSubject;
+    xCertificateTemplate.xSubject.ulValueLen = strlen( ( const char * ) pxSubject );
+    xCertificateTemplate.xValue.type = CKA_VALUE;
+    xCertificateTemplate.xValue.pValue = ( CK_VOID_PTR ) pcCertificate;
```

Comment:
This should be the certificate in BER form not PEM. ( DER is a specific case of BER) corePKCS11 has a function to convert between the two forms, or you could roll your own by removing the header, footer, and any CR/LF characters in the pem object and base64 decoding the remaining characters.

Reply:
Addressed in this commit: https://github.com/FreeRTOS/FreeRTOS/pull/724/commits/f1e56bc13f39a9abb75eab4a728dade01ecb6fc7